sopel/plugins/rules.py

```diff
@@ -1301,3 +1360,123 @@ def parse(self, text):
+    callback has per URL in the IRC line.
+
+    Here is an example with an URL rule with the pattern
+    ``r'https://example\\.com/(.*)'``:
+
+    .. code-block:: irc
+
+        <user> https://example.com/test
+        <Bot> You triggered an URL callback, with the "/test" path
+        <user> and this URL is https://example.com/other can you get it?
+        <Bot> You triggered an URL callback, with the "/other" path
+
+    Like generic rules, URL callback rules are not triggered by any specific
+    name and they don't have aliases.
+
+    .. note::
+
+        Unlike generic rules and commands, the :func:`~sopel.plugin.url`
+        decorator expect its decorated function to have the bot and the trigger
+        with a third parameter: the ``match`` parameter.
+
+        To use this class with an existing URL callback handler, the
+        :meth:`from_callable` classmethod **must** be used: it will wrap the
+        handler to work as intented. In that case, the ``trigger`` and the
+        ``match`` arguments will be the same when the rule executes.
+
+        This behavior makes the ``match`` parameter obsolete.
+
+    """
+    @classmethod
+    def from_callable(cls, settings, handler):
+        url_regexes = getattr(handler, 'url_regex', []) or []
+        if not url_regexes:
+            raise RuntimeError(
+                'Invalid URL callback: %s' % handler)
+
+        kwargs = cls.kwargs_from_callable(handler)
+
+        @functools.wraps(handler)
+        def execute_handler(bot, trigger):
+            return handler(bot, trigger, match=trigger)
+
+        kwargs.update({
+            'handler': execute_handler,
+            'schemes': settings.core.auto_url_schemes,
+        })
+
+        return cls(url_regexes, **kwargs)
+
+    @classmethod
+    def from_callable_lazy(cls, settings, handler):
+        url_lazy_loader = getattr(handler, 'url_lazy_loader', [])
+        if not url_lazy_loader:
+            raise RuntimeError(
+                'Invalid lazy loader URL callback: %s' % handler)
+
+        regexes = tuple(url_lazy_loader(settings))
+        kwargs = cls.kwargs_from_callable(handler)
+        kwargs.update({
+            'handler': handler,
+            'schemes': settings.core.auto_url_schemes,
+        })
+
+        return cls(regexes, **kwargs)
+
+    def __init__(self,
+                 regexes,
+                 schemes=None,
+                 **kwargs):
+        super(URLCallback, self).__init__(regexes, **kwargs)
+        # prevent mutability of registered schemes
+        self._schemes = tuple(schemes or URL_DEFAULT_SCHEMES)
+
+    def match(self, bot, pretrigger):
+        """Match URL in a pretrigger according to the rule.
+
+        :param bot: Sopel instance
+        :type bot: :class:`sopel.bot.Sopel`
+        :param pretrigger: Line to match
+        :type pretrigger: :class:`sopel.trigger.PreTrigger`
+
+        This method looks for URL in the IRC line, and for each it yields
```

Comment:
```suggestion
        This method looks for URLs in the IRC line, and for each it yields
```

I _think_ this is what you meant.